### PR TITLE
add load error code and remove banner size check

### PR DESCRIPTION
**Pull request by junjunsong**

### CODE REVIEW

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java [hunk 1]
```diff
@@ -49,14 +49,21 @@ public class MintegralConstants {
    */
   public static final int ERROR_CODE_NO_FILL = 104;
 
+  /**
+   * The returned Mintegral SDK init failed;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:53]: It makes more sense to update to below:
```
   * Mintegral SDK failed to initialize.
```

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java [hunk 2]
```diff
@@ -49,14 +49,21 @@ public class MintegralConstants {
...
+   */
+  public static final int ERROR_CODE_SDK_INIT_FAILED = 105;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:55]: Also to clarify, is this error only for initialization failure, or also for "Mintegral SDK isn't initialized"?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java [hunk 3]
```diff
@@ -49,14 +49,21 @@ public class MintegralConstants {
...
   @NonNull
   public static AdError createAdapterError(@AdapterError int errorCode,
       @NonNull String errorMessage) {
     return new AdError(errorCode, errorMessage, ERROR_DOMAIN);
   }
 
   @NonNull
-  public static AdError createSdkError(@NonNull String errorMessage) {
-    return new AdError(ERROR_MINTEGRAL_SDK, errorMessage, MINTEGRAL_SDK_ERROR_DOMAIN);
+  public static AdError createSdkError(int errorCode,@NonNull String errorMessage) {
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:66]: Nit, add space after ",".

Also, is there an interface for the Mintegral SDK error code?
If there is, we can add the annotation similar to above `@AdapterError int errorCode`.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java [hunk 4]
```diff
@@ -49,14 +49,21 @@ public class MintegralConstants {
...
+
   @NonNull
   public static AdError createAdapterError(@AdapterError int errorCode,
       @NonNull String errorMessage) {
     return new AdError(errorCode, errorMessage, ERROR_DOMAIN);
   }
 
   @NonNull
-  public static AdError createSdkError(@NonNull String errorMessage) {
-    return new AdError(ERROR_MINTEGRAL_SDK, errorMessage, MINTEGRAL_SDK_ERROR_DOMAIN);
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:60]: Is it guaranteed that Mintegral SDK return error code for all failure callbacks?

If Mintegral SDK still return string (without error code) for some of them, we should keep: 
``` 
public static AdError createSdkError(@NonNull String errorMessage) {
```
and create a new one for error code.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java [hunk 5]
```diff
@@ -44,45 +40,15 @@ public MintegralBannerAd(@NonNull MediationBannerAdConfiguration mediationBanner
...
-    if (closestSize.equals(AdSize.LEADERBOARD)) { // 728 * 90
-      bannerSize = new BannerSize(BannerSize.SMART_TYPE, closestSize.getWidth(), 0);
-    }
-    if (bannerSize == null) {
-      bannerSize = new BannerSize(BannerSize.DEV_SET_TYPE, closestSize.getWidth(),
-          closestSize.getHeight());
-    }
-    return bannerSize;
-  }
-
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java:76]: Can you clarify why is this removed? We would need to convert AdMob banner ad size to Mintegral banner ad size.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+    mintegralNativeAd.mapNativeAd(list.get(0));
+    nativeCallback = adLoadCallback.onSuccess(mintegralNativeAd);
+  }
+
+
+
+  @Override
+  public void onAdLoadErrorWithCode(int errorCode, String errorMessage) {
+
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:49]: Nit, remove empty space.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/rtb/MintegralRtbBannerAd.java [hunk 7]
```diff
@@ -29,20 +29,15 @@ public MintegralRtbBannerAd(
...
-          String.format("The requested banner size: %s is not supported by Mintegral SDK.",
-              adConfiguration.getAdSize()));
-      Log.e(TAG, bannerSizeError.toString());
-      adLoadCallback.onFailure(bannerSizeError);
-      return;
-    }
+    AdSize closestSize = adConfiguration.getAdSize();
+    BannerSize bannerSize = new BannerSize(BannerSize.DEV_SET_TYPE,
+            closestSize.getWidthInPixels(adConfiguration.getContext()),
+            closestSize.getHeightInPixels(adConfiguration.getContext()));
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/rtb/MintegralRtbBannerAd.java:35]: Since the `getMintegralBannerSizeFromAdMobAdSize()` method is removed, can you verify what banner ad sizes does Mintegral SDK support?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.android.gms.ads.mediation.MediationAdLoadCallback;
+import com.google.android.gms.ads.mediation.MediationNativeAdCallback;
+import com.google.android.gms.ads.mediation.UnifiedNativeAdMapper;
+import com.mbridge.msdk.out.Campaign;
+import com.mbridge.msdk.out.Frame;
+import com.mbridge.msdk.out.NativeAdWithCodeListener;
+
+import java.util.List;
+
+public class NativeAdWithCodeListenerImpl extends NativeAdWithCodeListener {
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:20]: We can have `NativeAdWithCodeListener` extends `MintegralNativeAd`? I don't see there is a need to create a new class for that.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java (hunk 1 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:53]: done

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java (hunk 2 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:55]: only for initialization failure

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java (hunk 3 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:66]: done

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java (hunk 4 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:60]: We have already processed all the internal processing in SDK, and some unexpected situations will be handled in mintegral

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java (hunk 5 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java:76]: After our internal communication, the size will be filtered during the server bid

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java (hunk 8 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:20]: I don't understand what you mean. Can you describe it in more detail?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java (hunk 1 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:53]: Seems like the change isn't applied yet. Can you make sure the latest commit is pushed?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java (hunk 3 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralConstants.java:66]: Ditto, seems like the change isn't applied yet. Can you make sure the latest commit is pushed?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralMediationAdapter.java [hunk 9]
```diff
@@ -172,7 +172,7 @@ public void onInitSuccess() {
 
       @Override
       public void onInitFail(String errorMessage) {
-        AdError initError = createSdkError(errorMessage);
+        AdError initError = createSdkError(MintegralConstants.ERROR_CODE_SDK_INIT_FAILED,errorMessage);
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralMediationAdapter.java:175]: Nit nit, add space after `,`.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java (hunk 8 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:20]: @junjunsong We can do `MintegralNativeAd extends NativeAdWithCodeListener` and have the event callbacks implemented in `MintegralNativeAd`, so we don't need an additional `NativeAdWithCodeListenerImpl` class.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java (hunk 6 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:49]: Not done yet, please make sure the latest commit is pushed.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/rtb/MintegralRtbBannerAd.java (hunk 7 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/rtb/MintegralRtbBannerAd.java:35]: Ditto from another comment, we should still keep this check as we shouldn't just rely on Mintegral backend implementation.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java (hunk 5 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java:76]: If the size of the adapter is checked, it may lead to a problem where the material meets the requirements but cannot be displayed. Or this location is mainly used to check a rough range rather than a strict standard, such as not less than 40px

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java (hunk 8 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:20]: if `MintegralNativeAd extends NativeAdWithCodeListener` , `MintegralNativeAd`  can't extends `UnifiedNativeAdMapper `, What do I need to do with `UnifiedNativeAdMapper `？

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralMediationAdapter.java (hunk 9 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/MintegralMediationAdapter.java:189]: @junjunsong Seems like this isn't addressed yet?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java (hunk 5 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java:87]: @junjunsong We need to keep this check. The `MediationUtils.findClosestSize()` method will return the banner ad size as long as the input ad size is within certain threshold (of the supported ad sizes). Therefore, the method will return an ad size as long as the publisher's input ad size is in the acceptable range. It will cover cases like adaptive banner.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java [hunk 10]
```diff
@@ -60,8 +61,8 @@ public void onResourceLoadSuccess(MBridgeIds mBridgeIds) {
   }
 
   @Override
-  public void onResourceLoadFail(MBridgeIds mBridgeIds, String errorMessage) {
-    AdError error = MintegralConstants.createSdkError(errorMessage);
+  public void onResourceLoadFailWithCode(MBridgeIds mBridgeIds, int errorCode, String errorMessage) {
+    AdError error = MintegralConstants.createSdkError(errorCode,errorMessage);
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java:65]: Nit, add space after `errorCode,`.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java [hunk 11]
```diff
@@ -82,9 +83,9 @@ public void onAdClose(MBridgeIds mBridgeIds, RewardInfo rewardInfo) {
   }
 
   @Override
-  public void onShowFail(MBridgeIds mBridgeIds, String errorMessage) {
-    AdError error = MintegralConstants.createAdapterError(MintegralConstants.ERROR_MINTEGRAL_SDK,
-        errorMessage);
+  public void onShowFailWithCode(MBridgeIds mBridgeIds, int errorCode, String errorMessage) {
+    AdError error = MintegralConstants.createSdkError(errorCode,
+            errorMessage);
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java:88]: Nit, these two lines should be able to fit in one (and number of chars still < 100).

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralRewardedAd.java [hunk 12]
```diff
@@ -59,8 +60,8 @@ public void onLoadSuccess(MBridgeIds mBridgeIds) {
   }
 
   @Override
-  public void onVideoLoadFail(MBridgeIds mBridgeIds, String errorMessage) {
-    AdError error = MintegralConstants.createSdkError(errorMessage);
+  public void onVideoLoadFailWithCode(MBridgeIds mBridgeIds, int errorCode, String errorMessage) {
+    AdError error = MintegralConstants.createSdkError(errorCode,errorMessage);
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralRewardedAd.java:64]: Nit, add space after `errorCode,`.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralRewardedAd.java [hunk 13]
```diff
@@ -106,9 +107,9 @@ public int getAmount() {
   }
 
   @Override
-  public void onShowFail(MBridgeIds mBridgeIds, String errorMessage) {
-    AdError error = MintegralConstants.createAdapterError(MintegralConstants.ERROR_MINTEGRAL_SDK,
-        errorMessage);
+  public void onShowFailWithCode(MBridgeIds mBridgeIds, int errorCode, String errorMessage) {
+    AdError error = MintegralConstants.createSdkError(errorCode,
+            errorMessage);
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralRewardedAd.java:112]: Nit, these two lines should be able to fit in one (and number of chars still < 100).

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/rtb/MintegralRtbBannerAd.java (hunk 7 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/rtb/MintegralRtbBannerAd.java:45]: @junjunsong We should revert this change and keep the check, as we can't assume publishers requesting banner ad with correct sizes all the time.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java (hunk 8 above)
Comment (joshh-devrel) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:20]: Oh my bad, `NativeAdWithCodeListener` is a class now.

Can you rename this class to: `MintegralNativeAdListener`?

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/waterfall/MintegralWaterfallBannerAd.java [hunk 14]
```diff
@@ -42,17 +42,10 @@ public MintegralWaterfallBannerAd(
...
-              adConfiguration.getAdSize()));
-      Log.e(TAG, bannerSizeError.toString());
-      adLoadCallback.onFailure(bannerSizeError);
-      return;
-    }
-
+    AdSize closestSize = adConfiguration.getAdSize();
+    BannerSize bannerSize = new BannerSize(BannerSize.DEV_SET_TYPE,
+            closestSize.getWidthInPixels(adConfiguration.getContext()),
+            closestSize.getHeightInPixels(adConfiguration.getContext()));
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/waterfall/MintegralWaterfallBannerAd.java:48]: Ditto, please feel free to clarify the need to remove the check here.

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java (hunk 5 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java:87]: The banner of mintgeneral can adapt to any size of camp. Are you worried about the size set by the developer, and the mintegral sdk won't display properly in certain sizes?

Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralBannerAd.java:87]: In other words, as long as our server returns the material, the client can adapt the material well

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralRewardedAd.java (hunk 12 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralRewardedAd.java:64]: done

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java (hunk 11 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java:88]: done

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java (hunk 10 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/MintegralInterstitialAd.java:65]: done

File: ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java (hunk 8 above)
Comment (junjunsong) [reply ThirdPartyAdapters/mintegral/mintegral/src/main/java/com/google/ads/mediation/mintegral/mediation/NativeAdWithCodeListenerImpl.java:20]: done
